File: scripts/recover_completed_runs.R
#!/usr/bin/env Rscript
# Recover model runs that completed in Plumber but are marked failed/running in the DB.
# Happens when plumber-sync status probe got a transient 500 during Plumber auth-crash window.
#
# Usage:
#   ./scripts/recover_completed_runs.R            # real run (inside plumber container)
#   ./scripts/recover_completed_runs.R --dry-run   # preview only
#
# Must run inside the Plumber container where jsonlite is available, OR
# pass JOBS_DIR=/path/to/jobs and DATABASE_URL to run from host.

dry_run <- "--dry-run" %in% commandArgs(TRUE)

# --- Determine execution mode ---
IN_CONTAINER <- dir.exists("/app/R")
if (IN_CONTAINER) {
  jobs_dir <- "/app/outputs/jobs"
  psql_cmd <- function(sql) {
    system2("psql", c("-U", "sdm", "-d", "sdm_platform", "-c", sql),
      stdout = TRUE, stderr = TRUE)
  }
  # jsonlite IS available inside the container
  library(jsonlite)
  parse_json <- function(path) {
    fromJSON(path, simplifyVector = FALSE)
  }
  to_json <- function(x) {
    toJSON(x, auto_unbox = TRUE, null = "null")
  }
} else {
  jobs_dir <- Sys.getenv("JOBS_DIR", file.path("outputs", "jobs"))
  psql_cmd <- function(sql) {
    system2("docker", c(
      "exec", "-i", "sdm-dashboard-main-postgres-1",
      "psql", "-U", "sdm", "-d", "sdm_platform",
      "-c", sql
    ), stdout = TRUE, stderr = TRUE)
  }
  # Use python3 for JSON (no jsonlite dependency on host)
  parse_json <- function(path) {
    json_str <- system2("python3", c("-c", sprintf(
      "import json,sys; print(json.dumps(json.load(open('%s'))))", path
    )), stdout = TRUE, stderr = FALSE)
    if (length(json_str) == 0) return(NULL)
    jsonlite::fromJSON(json_str, simplifyVector = FALSE)
  }
  to_json <- function(x) {
    json_str <- system2("python3", c("-c", sprintf(
      "import json,sys; print(json.dumps(%s))",
      deparse(x)
    )), stdout = TRUE, stderr = FALSE)
    json_str
  }
  # Fallback: try loading jsonlite anyway
  if (!requireNamespace("jsonlite", quietly = TRUE)) {
    # Use a minimal R-native replacement
    parse_json <- function(path) {
      raw <- readLines(path, warn = FALSE)
      if (length(raw) == 0) return(NULL)
      # crude list-of-lists only; works for our simple meta.json structure
      eval(parse(text = paste0("list(", gsub('"', "'", paste(raw, collapse = "\n")), ")")))
    }
    to_json <- function(x) {
      jsonlite::toJSON(x, auto_unbox = TRUE, null = "null")
    }
  } else {
    library(jsonlite)
    parse_json <- function(path) fromJSON(path, simplifyVector = FALSE)
    to_json <- function(x) toJSON(x, auto_unbox = TRUE, null = "null")
  }
}

if (!dir.exists(jobs_dir)) stop("jobs dir not found: ", jobs_dir, call. = FALSE)

job_dirs <- list.dirs(jobs_dir, full.names = TRUE, recursive = FALSE)
cat(sprintf("Scanning %d job directories...\n", length(job_dirs)))

recovered <- 0L
skipped <- 0L

for (jd in job_dirs) {
  meta_file <- file.path(jd, "meta.json")
  if (!file.exists(meta_file)) {
    skipped <- skipped + 1L
    next
  }

  meta <- tryCatch(parse_json(meta_file), error = function(e) NULL)
  if (is.null(meta)) {
    skipped <- skipped + 1L
    next
  }

  if (!identical(meta$status, "completed")) next

  # Plumber writes config.runId matching the DB runs.id
  run_id <- meta$config$runId
  if (is.null(run_id) || !nchar(as.character(run_id))) {
    skipped <- skipped + 1L
    next
  }

  cat(sprintf("\n  Job: %s  RunId: %s  Completed: %s\n",
    basename(jd), run_id,
    sub("T", " ", meta$completed_at %||% "?")))

  if (dry_run) {
    cat("    DRY RUN — would update DB\n")
    next
  }

  # Build metrics/output_files as JSON strings for psql
  metrics_json <- if (!is.null(meta$metrics)) {
    gsub("'", "''", to_json(meta$metrics))
  } else "NULL"

  output_files_json <- if (!is.null(meta$output_files)) {
    gsub("'", "''", to_json(meta$output_files))
  } else "NULL"

  completed_at <- sub("T", " ", meta$completed_at %||% Sys.time())

  sql <- sprintf(
    "UPDATE runs SET status = 'completed', error = NULL, error_code = NULL, error_hint = NULL, metrics = '%s'::jsonb, output_files = '%s'::jsonb, completed_at = '%s'::timestamp WHERE id = '%s' AND status IN ('failed', 'running');",
    metrics_json, output_files_json, completed_at, run_id
  )

  result <- psql_cmd(sql)

  if (any(grepl("UPDATE 1", result))) {
    cat("    RECOVERED — DB updated to completed\n")
    recovered <- recovered + 1L
  } else if (any(grepl("UPDATE 0", result))) {
    cat("    SKIPPED — run not in failed/running state, or id not found\n")
  } else {
    cat("    ERROR —", paste(result, collapse = "\n"), "\n")
  }
}

cat(sprintf("\nDone. Recovered: %d  Skipped/irrelevant: %d\n", recovered, skipped))

File: scripts/recover_completed_runs.py
#!/usr/bin/env python3
"""Recover runs that completed in Plumber but are marked failed/running in the DB.

Usage:
  python3 scripts/recover_completed_runs.py            # real run
  python3 scripts/recover_completed_runs.py --dry-run   # preview only
"""
import json
import os
import subprocess
import sys

dry_run = "--dry-run" in sys.argv

JOBS_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "outputs", "jobs")
PSQL = ["docker", "exec", "-i", "sdm-dashboard-main-postgres-1", "psql", "-U", "sdm", "-d", "sdm_platform"]


def psql(sql):
    result = subprocess.run(PSQL + ["-c", sql], capture_output=True, text=True)
    return result.stdout, result.stderr


def esc(val):
    """Escape a string for use inside a psql -c single-quoted SQL string."""
    if val is None:
        return "NULL"
    return "'" + val.replace("'", "''") + "'"


recovered = 0
skipped = 0

with os.scandir(JOBS_DIR) as it:
    for entry in it:
        if not entry.is_dir():
            continue
        meta_file = os.path.join(entry.path, "meta.json")
        if not os.path.isfile(meta_file):
            skipped += 1
            continue

        try:
            with open(meta_file) as f:
                meta = json.load(f)
        except (json.JSONDecodeError, OSError):
            skipped += 1
            continue

        if meta.get("status") != "completed":
            continue

        run_id = meta.get("config", {}).get("runId")
        if not run_id or not isinstance(run_id, str) or not run_id.strip():
            skipped += 1
            continue

        completed_at = meta.get("completed_at", "").replace("T", " ").replace("Z", "")
        if not completed_at:
            skipped += 1
            continue

        metrics = meta.get("metrics")
        output_files = meta.get("output_files")

        metrics_json = esc(json.dumps(metrics, ensure_ascii=False)) if metrics else "NULL"
        output_files_json = esc(json.dumps(output_files, ensure_ascii=False)) if output_files else "NULL"
        completed_at_esc = esc(completed_at)

        if dry_run:
            print(f"  Job: {entry.name}  RunId: {run_id}  Would recover from 'failed'/'running' → 'completed'")
            continue

        sql = (
            f"UPDATE runs SET "
            f"status = 'completed', "
            f"error = NULL, "
            f"error_code = NULL, "
            f"error_hint = NULL, "
            f"metrics = {metrics_json}::jsonb, "
            f"output_files = {output_files_json}::jsonb, "
            f"completed_at = {completed_at_esc}::timestamp "
            f"WHERE id = {esc(run_id)} AND status IN ('failed', 'running');"
        )

        stdout, stderr = psql(sql)

        if "UPDATE 1" in stdout:
            print(f"  RECOVERED: {entry.name} → RunId: {run_id}")
            recovered += 1
        elif "UPDATE 0" in stdout:
            print(f"  SKIPPED: {entry.name} (not in failed/running state or id not found)")
        else:
            print(f"  ERROR: {entry.name} — {stdout.strip()} {stderr.strip()}")


print(f"\nDone. Recovered: {recovered}  Skipped/irrelevant: {skipped}")
